### PR TITLE
Uri support

### PR DESCRIPTION
**Pull request by jszczygiel**

### CODE REVIEW

**Comment by Piasy** [comment]
Please make it compile, and note about the space between parameters.

**Comment by jszczygiel** [comment]
Hello,

It seems that there is some problem with ANDROID_NDK_HOME path variable. I've not touch any native code and I don't seem to know where that problem might come

**Comment by Piasy** [comment]
You didn't even open it in IDE, or run a gradle build task? You need import the `Uri` class.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/PlayConfig.java [hunk 1]
```diff
@@ -74,6 +78,14 @@ public static Builder res(Context context, @RawRes int audioResource) {
         return builder;
     }
 
+    public static Builder uri(Context context, Uri uri){
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/PlayConfig.java]: space between `){` please.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 2]
```diff
@@ -58,6 +58,29 @@ public static RxAudioPlayer getInstance() {
...
         }
         switch (config.mType) {
+            case PlayConfig.TYPE_URI:
+                return Observable.create(emitter -> {
+                    stopPlay();
+
+                    Log.d(TAG, "MediaPlayer to start play: " + config.mUri);
+                    mPlayer = new MediaPlayer();
+                    try {
+                        mPlayer.setDataSource(config.mContext,config.mUri);
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: space please.

File: rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java [hunk 3]
```diff
@@ -199,6 +222,23 @@ public boolean playNonRxy(@NonNull final PlayConfig config,
             return false;
         }
         switch (config.mType) {
+            case PlayConfig.TYPE_URI:
+                Log.d(TAG, "MediaPlayer to start play: " + config.mUrl);
+                mPlayer = new MediaPlayer();
+                try {
+                    mPlayer.setDataSource(config.mContext,config.mUri);
```
Comment (Piasy) [inline rxandroidaudio/src/main/java/com/github/piasy/rxandroidaudio/RxAudioPlayer.java]: space please.

**Review by Piasy — CHANGES_REQUESTED** [review]

**Comment by jszczygiel** [comment]
```
szczygiel@szczygiel ~/Downloads/RxAndroidAudio-uri_support: ./gradlew assembleDebug
extractDebugAnnotations is incompatible with java 8 sources and has been disabled.
extractReleaseAnnotations is incompatible with java 8 sources and has been disabled.
Jack is disabled, but one of the plugins you are using supports Java 8 language features.
Jack is disabled, but one of the plugins you are using supports Java 8 language features.
:AudioProcessor:preBuild UP-TO-DATE
:AudioProcessor:preDebugBuild UP-TO-DATE
:AudioProcessor:checkDebugManifest
:AudioProcessor:prepareDebugDependencies
:AudioProcessor:compileDebugAidl
:AudioProcessor:compileLint
:AudioProcessor:copyDebugLint NO-SOURCE
:AudioProcessor:mergeDebugShaders
:AudioProcessor:compileDebugShaders
:AudioProcessor:generateDebugAssets
:AudioProcessor:mergeDebugAssets
:AudioProcessor:mergeDebugProguardFiles UP-TO-DATE
:AudioProcessor:packageDebugRenderscript NO-SOURCE
:AudioProcessor:compileDebugRenderscript
:AudioProcessor:generateDebugResValues
:AudioProcessor:generateDebugResources
:AudioProcessor:packageDebugResources
:AudioProcessor:processDebugManifest
:AudioProcessor:generateDebugBuildConfig
:AudioProcessor:processDebugResources
:AudioProcessor:generateDebugSources
:AudioProcessor:incrementalDebugJavaCompilationSafeguard
:AudioProcessor:javaPreCompileDebug
:AudioProcessor:compileDebugJavaWithJavac
:AudioProcessor:compileDebugJavaWithJavac - is not incremental (e.g. outputs have changed, no previous execution, etc.).
:AudioProcessor:processDebugJavaRes NO-SOURCE
:AudioProcessor:transformResourcesWithMergeJavaResForDebug
:AudioProcessor:transformClassesAndResourcesWithSyncLibJarsForDebug
:AudioProcessor:generateJsonModelDebug UP-TO-DATE
:AudioProcessor:externalNativeBuildDebug
Build audio-processor mips64
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/mips64/libaudio-processor.so
Build audio-processor mips
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/mips/libaudio-processor.so
Build audio-processor x86_64
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/x86_64/libaudio-processor.so
Build audio-processor x86
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/x86/libaudio-processor.so
Build audio-processor arm64-v8a
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/arm64-v8a/libaudio-processor.so
Build audio-processor armeabi-v7a
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/armeabi-v7a/libaudio-processor.so
Build audio-processor armeabi
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/debug/obj/armeabi/libaudio-processor.so
:AudioProcessor:mergeDebugJniLibFolders
:AudioProcessor:transformNativeLibsWithMergeJniLibsForDebug
:AudioProcessor:transformNativeLibsWithStripDebugSymbolForDebug
:AudioProcessor:transformNativeLibsWithSyncJniLibsForDebug
:AudioProcessor:bundleDebug
:AudioProcessor:compileDebugSources
:AudioProcessor:assembleDebug
:AudioProcessor:preReleaseBuild UP-TO-DATE
:AudioProcessor:checkReleaseManifest
:AudioProcessor:prepareReleaseDependencies
:AudioProcessor:compileReleaseAidl
:AudioProcessor:copyReleaseLint NO-SOURCE
:AudioProcessor:mergeReleaseShaders
:AudioProcessor:compileReleaseShaders
:AudioProcessor:generateReleaseAssets
:AudioProcessor:mergeReleaseAssets
:AudioProcessor:mergeReleaseProguardFiles UP-TO-DATE
:AudioProcessor:packageReleaseRenderscript NO-SOURCE
:AudioProcessor:compileReleaseRenderscript
:AudioProcessor:generateReleaseResValues
:AudioProcessor:generateReleaseResources
:AudioProcessor:packageReleaseResources
:AudioProcessor:processReleaseManifest
:AudioProcessor:generateReleaseBuildConfig
:AudioProcessor:processReleaseResources
:AudioProcessor:generateReleaseSources
:AudioProcessor:incrementalReleaseJavaCompilationSafeguard
:AudioProcessor:javaPreCompileRelease
:AudioProcessor:compileReleaseJavaWithJavac
:AudioProcessor:compileReleaseJavaWithJavac - is not incremental (e.g. outputs have changed, no previous execution, etc.).
:AudioProcessor:processReleaseJavaRes NO-SOURCE
:AudioProcessor:transformResourcesWithMergeJavaResForRelease
:AudioProcessor:transformClassesAndResourcesWithSyncLibJarsForRelease
:AudioProcessor:generateJsonModelRelease UP-TO-DATE
:AudioProcessor:externalNativeBuildRelease
Build audio-processor mips64
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/mips64/libaudio-processor.so
Build audio-processor mips
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/mips/libaudio-processor.so
Build audio-processor x86_64
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/x86_64/libaudio-processor.so
Build audio-processor x86
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/x86/libaudio-processor.so
Build audio-processor arm64-v8a
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/arm64-v8a/libaudio-processor.so
Build audio-processor armeabi-v7a
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/armeabi-v7a/libaudio-processor.so
Build audio-processor armeabi
[1/2] Building C object CMakeFiles/audio-processor.dir/smbPitchShift.c.o
[2/2] Linking C shared library /usr/local/google/home/szczygiel/Downloads/RxAndroidAudio-uri_support/AudioProcessor/build/intermediates/cmake/release/obj/armeabi/libaudio-processor.so
:AudioProcessor:mergeReleaseJniLibFolders
:AudioProcessor:transformNativeLibsWithMergeJniLibsForRelease
:AudioProcessor:transformNativeLibsWithStripDebugSymbolForRelease
:AudioProcessor:transformNativeLibsWithSyncJniLibsForRelease
:AudioProcessor:bundleRelease
:app:preBuild UP-TO-DATE
:app:preDebugBuild UP-TO-DATE
:app:checkDebugManifest
:app:preReleaseBuild UP-TO-DATE
:app:prepareComAndroidSupportAnimatedVectorDrawable2520Library
:app:prepareComAndroidSupportAppcompatV72520Library
:app:prepareComAndroidSupportSupportCompat2520Library
:app:prepareComAndroidSupportSupportCoreUi2520Library
:app:prepareComAndroidSupportSupportCoreUtils2520Library
:app:prepareComAndroidSupportSupportFragment2520Library
:app:prepareComAndroidSupportSupportMediaCompat2520Library
:app:prepareComAndroidSupportSupportV42520Library
:app:prepareComAndroidSupportSupportVectorDrawable2520Library
:app:prepareComJakewhartonButterknife840Library
:app:prepareComTbruyelleRxpermissions2Rxpermissions091Library
:app:prepareComTrelloRxlifecycle2RxlifecycleAndroid201Library
:app:prepareComTrelloRxlifecycle2RxlifecycleComponents201Library
:app:prepareIoReactivexRxjava2Rxandroid201Library
:rxandroidaudio:preBuild UP-TO-DATE
:rxandroidaudio:preReleaseBuild UP-TO-DATE
:rxandroidaudio:checkReleaseManifest
:rxandroidaudio:prepareReleaseDependencies
:rxandroidaudio:compileReleaseAidl
:rxandroidaudio:compileReleaseNdk NO-SOURCE
:rxandroidaudio:compileLint
:rxandroidaudio:copyReleaseLint NO-SOURCE
:rxandroidaudio:compileReleaseRenderscript
:rxandroidaudio:generateReleaseBuildConfig
:rxandroidaudio:generateReleaseResValues
:rxandroidaudio:generateReleaseResources
:rxandroidaudio:packageReleaseResources
:rxandroidaudio:processReleaseManifest
:rxandroidaudio:processReleaseResources
:rxandroidaudio:generateReleaseSources
:rxandroidaudio:incrementalReleaseJavaCompilationSafeguard
:rxandroidaudio:javaPreCompileRelease
:rxandroidaudio:compileReleaseJavaWithJavac
:rxandroidaudio:extractReleaseAnnotations
:rxandroidaudio:mergeReleaseShaders
:rxandroidaudio:compileReleaseShaders
:rxandroidaudio:generateReleaseAssets
:rxandroidaudio:mergeReleaseAssets
:rxandroidaudio:mergeReleaseProguardFiles UP-TO-DATE
:rxandroidaudio:packageReleaseRenderscript NO-SOURCE
:rxandroidaudio:transformClassesWithRetrolambdaForRelease
:rxandroidaudio:processReleaseJavaRes NO-SOURCE
:rxandroidaudio:transformResourcesWithMergeJavaResForRelease
:rxandroidaudio:transformClassesAndResourcesWithSyncLibJarsForRelease
:rxandroidaudio:mergeReleaseJniLibFolders
:rxandroidaudio:transformNativeLibsWithMergeJniLibsForRelease
:rxandroidaudio:transformNativeLibsWithStripDebugSymbolForRelease
:rxandroidaudio:transformNativeLibsWithSyncJniLibsForRelease
:rxandroidaudio:bundleRelease
:app:prepareDebugDependencies
:app:compileDebugAidl
:app:compileDebugRenderscript
:app:generateDebugBuildConfig
:app:generateDebugResValues
:app:generateDebugResources
:app:mergeDebugResources
:app:processDebugManifest
:app:processDebugResources
:app:generateDebugSources
:app:incrementalDebugJavaCompilationSafeguard
:app:javaPreCompileDebug
:app:compileDebugJavaWithJavac
:app:compileDebugNdk NO-SOURCE
:app:compileDebugSources
:app:mergeDebugShaders
:app:compileDebugShaders
:app:generateDebugAssets
:app:mergeDebugAssets
:app:transformClassesWithRetrolambdaForDebug
:app:transformClassesWithDexForDebug

Running dex as a separate process.

To run dex in process, the Gradle daemon needs a larger heap.
It currently has 1024 MB.
For faster builds, increase the maximum heap size for the Gradle daemon to at least 1536 MB.
To do this set org.gradle.jvmargs=-Xmx1536M in the project gradle.properties.
For more information see https://docs.gradle.org/current/userguide/build_environment.html

:app:mergeDebugJniLibFolders
:app:transformNativeLibsWithMergeJniLibsForDebug
:app:transformNativeLibsWithStripDebugSymbolForDebug
:app:processDebugJavaRes NO-SOURCE
:app:transformResourcesWithMergeJavaResForDebug
:app:validateSigningDebug
:app:packageDebug
:app:assembleDebug
:rxandroidaudio:preDebugBuild UP-TO-DATE
:rxandroidaudio:checkDebugManifest
:rxandroidaudio:prepareDebugDependencies
:rxandroidaudio:compileDebugAidl
:rxandroidaudio:compileDebugNdk NO-SOURCE
:rxandroidaudio:copyDebugLint NO-SOURCE
:rxandroidaudio:compileDebugRenderscript
:rxandroidaudio:generateDebugBuildConfig
:rxandroidaudio:generateDebugResValues
:rxandroidaudio:generateDebugResources
:rxandroidaudio:packageDebugResources
:rxandroidaudio:processDebugManifest
:rxandroidaudio:processDebugResources
:rxandroidaudio:generateDebugSources
:rxandroidaudio:incrementalDebugJavaCompilationSafeguard
:rxandroidaudio:javaPreCompileDebug
:rxandroidaudio:compileDebugJavaWithJavac
:rxandroidaudio:extractDebugAnnotations
:rxandroidaudio:mergeDebugShaders
:rxandroidaudio:compileDebugShaders
:rxandroidaudio:generateDebugAssets
:rxandroidaudio:mergeDebugAssets
:rxandroidaudio:mergeDebugProguardFiles UP-TO-DATE
:rxandroidaudio:packageDebugRenderscript NO-SOURCE
:rxandroidaudio:transformClassesWithRetrolambdaForDebug
:rxandroidaudio:processDebugJavaRes NO-SOURCE
:rxandroidaudio:transformResourcesWithMergeJavaResForDebug
:rxandroidaudio:transformClassesAndResourcesWithSyncLibJarsForDebug
:rxandroidaudio:mergeDebugJniLibFolders
:rxandroidaudio:transformNativeLibsWithMergeJniLibsForDebug
:rxandroidaudio:transformNativeLibsWithStripDebugSymbolForDebug
:rxandroidaudio:transformNativeLibsWithSyncJniLibsForDebug
:rxandroidaudio:bundleDebug
:rxandroidaudio:compileDebugSources
:rxandroidaudio:assembleDebug

BUILD SUCCESSFUL

Total time: 7.349 secs
```

Build runs smoothly on my machine.

Well I run it in Andorid studio. Uri import is not needed in RxAudioPlayer, but it is required in PlayConfig, which was satisfied. This is due to not using type explicitly in RxAudioPlayer.

**Review by jszczygiel — COMMENTED** [review]
done

**Pull request merged**